File: backend/app/core/telemetry.py
"""PR3/C3.1 — backend telemetry (PostHog).

Single chokepoint for all server-side product analytics.

Design rules:

  1. **No-op safe.** When `POSTHOG_KEY` is unset or `posthog-node` is
     not installed, every public function silently does nothing. Dev,
     CI, and self-hosted deployments without telemetry must work
     identically to a production deploy with PostHog wired up — they
     just stop emitting events. We never raise from here.

  2. **One init.** A module-level `_client` initialized on first use,
     guarded by an `_initialized` flag. Workers that fork (Celery)
     re-init on first event in the worker process; the SDK handles
     this safely.

  3. **Thin wrapper.** We intentionally do not add a typed event
     catalog at this layer — that lives in the call sites in PR3/C3.2.
     This module just gives us `capture(distinct_id, event, properties)`
     and `flush()`.

  4. **Async-friendly.** PostHog's Python SDK is sync; calls are fire-
     and-forget (it has its own background queue). We expose `capture`
     as a normal `def` and let callers invoke it from inside async
     handlers without `await`.
"""

from __future__ import annotations

import os
from typing import Any

import structlog

log = structlog.get_logger()

# Module-level singleton. None means either not configured yet, or
# permanently disabled (no key set).
_client: Any | None = None
_initialized: bool = False


def _maybe_init() -> Any | None:
    """Lazy init. Returns the live client or None if telemetry is off."""
    global _client, _initialized

    if _initialized:
        return _client

    _initialized = True
    api_key = os.environ.get("POSTHOG_KEY")
    if not api_key:
        log.info("telemetry.posthog_disabled", reason="POSTHOG_KEY not set")
        return None

    try:
        # Imported lazily so an optional dep doesn't crash the boot.
        from posthog import Posthog  # type: ignore[import-not-found]
    except ImportError:
        log.warning(
            "telemetry.posthog_disabled",
            reason="posthog package not installed",
        )
        return None

    host = os.environ.get("POSTHOG_HOST", "https://app.posthog.com")
    try:
        _client = Posthog(api_key, host=host)
        log.info("telemetry.posthog_enabled", host=host)
    except Exception as exc:
        log.warning("telemetry.posthog_init_failed", error=str(exc))
        _client = None

    return _client


def capture(
    distinct_id: str | None,
    event: str,
    properties: dict[str, Any] | None = None,
) -> None:
    """Fire a PostHog event. No-op if telemetry is disabled.

    Args:
        distinct_id: The user's UUID (str) or None for anonymous events.
                     Anonymous events get bucketed by a literal "anon"
                     id so server-side aggregations still see them.
        event:       Snake_case event name (e.g. "today.warmup_done").
        properties:  Arbitrary structured props attached to the event.

    We swallow every exception from the SDK so a PostHog outage can
    never take down a request handler.
    """
    client = _maybe_init()
    if client is None:
        return

    try:
        client.capture(
            distinct_id=distinct_id or "anon",
            event=event,
            properties=properties or {},
        )
    except Exception as exc:
        # Soft fail — telemetry is never load-bearing.
        # NB: structlog reserves the positional first arg as `event`,
        # so we rename our kwarg to `event_name` to avoid the
        # `multiple values for argument 'event'` collision.
        log.warning(
            "telemetry.capture_failed",
            event_name=event,
            error=str(exc),
        )


def flush() -> None:
    """Force the SDK's background queue to drain. Useful at app
    shutdown so events from the last request actually leave the
    process. No-op if telemetry is disabled."""
    client = _maybe_init()
    if client is None:
        return
    try:
        client.flush()
    except Exception as exc:
        log.warning("telemetry.flush_failed", error=str(exc))

File: frontend/src/lib/telemetry.ts
/**
 * PR3/C3.1 — frontend telemetry (PostHog).
 *
 * Single chokepoint for all client-side product analytics.
 *
 * Design rules:
 *
 *   1. **No-op safe.** When `NEXT_PUBLIC_POSTHOG_KEY` is unset, every
 *      function silently does nothing. Dev, CI, and self-hosted
 *      deployments without telemetry must work identically to a
 *      production deploy with PostHog wired up — they just stop
 *      emitting events. We never throw from here.
 *
 *   2. **Lazy import.** `posthog-js` only gets pulled in when we
 *      actually have a key. This keeps the dev-server bundle smaller
 *      for engineers who never run with telemetry on.
 *
 *   3. **Single init.** Module-scoped guard prevents double-init when
 *      Next's HMR re-evaluates the module in dev. Calling identify()
 *      after init is allowed and re-runs every time the user changes.
 *
 *   4. **Thin wrapper.** No typed event catalog at this layer — the
 *      catalog lives at the call sites in PR3/C3.2 so a screen-level
 *      refactor can rename or drop events without touching this file.
 */

type PostHogClient = {
  init: (key: string, opts: Record<string, unknown>) => void;
  capture: (event: string, properties?: Record<string, unknown>) => void;
  identify: (
    distinctId: string,
    properties?: Record<string, unknown>,
  ) => void;
  reset: () => void;
};

let _client: PostHogClient | null = null;
let _initStarted = false;

function getKey(): string | null {
  // process.env.NEXT_PUBLIC_* is statically inlined at build time by
  // Next, so this works in both client and server bundles.
  const key = process.env.NEXT_PUBLIC_POSTHOG_KEY;
  return key && key.length > 0 ? key : null;
}

async function ensureClient(): Promise<PostHogClient | null> {
  if (_client) return _client;
  if (_initStarted) return _client; // race-safe: another caller is mid-init
  _initStarted = true;

  if (typeof window === "undefined") {
    // Server-side rendering path; PostHog browser SDK is no-op here.
    return null;
  }

  const key = getKey();
  if (!key) return null;

  try {
    const mod = (await import("posthog-js")) as unknown as {
      default: PostHogClient;
    };
    const ph = mod.default;
    ph.init(key, {
      api_host:
        process.env.NEXT_PUBLIC_POSTHOG_HOST ?? "https://app.posthog.com",
      // Don't auto-capture — we want a curated event catalog (C3.2),
      // not a noisy click-stream that bloats the workspace.
      autocapture: false,
      capture_pageview: false,
      // Respect user privacy by default.
      persistence: "localStorage",
    });
    _client = ph;
    return _client;
  } catch {
    // Soft fail — telemetry is never load-bearing.
    return null;
  }
}

/**
 * Fire a PostHog event. Fire-and-forget — we don't await the SDK init
 * because callers (button onClick, mutation onSuccess) shouldn't block
 * on telemetry network round-trips.
 */
export function capture(
  event: string,
  properties?: Record<string, unknown>,
): void {
  if (!getKey() || typeof window === "undefined") return;
  void (async () => {
    const client = await ensureClient();
    if (!client) return;
    try {
      client.capture(event, properties);
    } catch {
      // Soft fail.
    }
  })();
}

/**
 * Tag the current session with the signed-in user's id so server-side
 * and client-side events share a distinct_id.
 */
export function identify(
  distinctId: string,
  properties?: Record<string, unknown>,
): void {
  if (!getKey() || typeof window === "undefined") return;
  void (async () => {
    const client = await ensureClient();
    if (!client) return;
    try {
      client.identify(distinctId, properties);
    } catch {
      // Soft fail.
    }
  })();
}

/** Drop all session state on sign-out. */
export function reset(): void {
  if (!_client) return;
  try {
    _client.reset();
  } catch {
    // Soft fail.
  }
}
